### PR TITLE
refactor: Add labeled blocks with goto statement

### PR DESCRIPTION
**Pull request by tristanmenzel**
## Proposed Changes

- Add label property to blocks so they can be 'referenced' by name
- Add goto node for unconditional navigation to a labelled block
-

### CODE REVIEW

File: src/puya/ir/builder/blocks.py [hunk 1]
```diff
@@ -164,3 +166,30 @@ def loop_continue(self, source_location: SourceLocation) -> None:
...
+        for block in self._labelled_blocks.values():
+            self.ssa.seal_block(block)
+
+    def goto_label(self, label: str, source_location: SourceLocation) -> None:
+        try:
+            target = self._labelled_blocks[label]
+            self.goto(target, source_location)
+        except KeyError:
+            target = self._labelled_blocks[label] = BasicBlock(
+                source_location=source_location, comment="Temp"
```
Comment (tristanmenzel) [inline src/puya/ir/builder/blocks.py:193]: Can't use the correct source_location or comment here as it's not known at the time. We could 'replace' the block once it is known?

File: src/puya/ir/builder/blocks.py [hunk 2]
```diff
@@ -164,3 +166,30 @@ def loop_continue(self, source_location: SourceLocation) -> None:
...
+    def make_labelled_block(
+        self, label: str, source_location: SourceLocation, description: str | None
+    ) -> BasicBlock:
+        try:
+            new_block = self._labelled_blocks[label]
+        except KeyError:
+            new_block = self._labelled_blocks[label] = BasicBlock(
+                source_location=source_location, comment=description
+            )
+        self.goto(new_block)
```
Comment (daniel-makerx) [inline src/puya/ir/builder/blocks.py:179]: could add an else block here and set source_location and comment on the new_block to address the issue you mentioned below?

File: src/puya/awst/function_traverser.py [hunk 3]
```diff
@@ -1,5 +1,6 @@
 import puya.awst.visitors
 from puya.awst import nodes as awst_nodes
+from puya.awst.visitors import T
```
Comment (daniel-makerx) [inline src/puya/awst/function_traverser.py:3]: Shouldn't need this?

File: src/puya/ir/builder/blocks.py (hunk 2 above)
Comment (tristanmenzel) [reply src/puya/ir/builder/blocks.py:179]: oh for some reason I thought ir nodes were immutable

File: src/puya/awst/function_traverser.py (hunk 3 above)
Comment (tristanmenzel) [reply src/puya/awst/function_traverser.py]: Just a remnant of pycharm auto creating the missing abstract member

**Comment by tristanmenzel** [comment]
closing in favour of #279

**Pull request closed without merging**